bananas_api/new_upload/readers/heightmap.py

```diff
@@ -39,6 +47,50 @@ def read(self, fp):
 
         try:
             im = Image.open(fp)
-            self.size = im.size
         except Exception:
             raise ValidationException("File is not a valid image.")
+
+        # Limit the size of heightmaps, as otherwise it could allocate a lot of
+        # memory for clients loading the map.
+        if im.width * im.height > 16384 * 16384:
+            raise ValidationException("Image is too large.")
+
+        self.size = im.size
+
+        # The following code is based on https://github.com/OpenTTD/OpenTTD/blob/master/src/heightmap.cpp
+        # to mimic the heightmap loading in OpenTTD as much as possible.
+
+        if im.palette:
+            palette = im.palette.palette
+            mode_len = len(im.palette.mode)
+
+            gray_palette = []
+            all_gray = True
+            for i in range(0, len(palette), mode_len):
+                color = palette[i : i + mode_len]
+                all_gray = all_gray and (color[0] == color[1] and color[1] == color[2])
+                gray_palette.append(rgb_to_gray(color))
+
+            palette_size = len(palette) // mode_len
+            if palette_size == 16 and not all_gray:
+                for i in range(palette_size):
+                    gray_palette[i] = 256 * i // palette_size
+
+        if im.mode == "P":
+            height_func = lambda pixel: gray_palette[pixel]
```

Comment:
if you make a dict mapping modes to lambdas instead, does it complain less? (you'd have to duplicate the RGB/RGBA one, but maybe that's ok?)

Reply:
Works! Nice idea :D